File: apps/ios/GalaxySSI/AgentFinalResponseIdentity.swift
import Foundation

enum AgentFinalResponseIdentity {
  static func dedupeKey(
    turnId: String,
    sourceMessageId: Int64 = 0,
    taskId: String = ""
  ) -> String {
    let identity: String
    if !isBlank(turnId) {
      identity = "turn:\(trim(turnId))"
    } else if sourceMessageId > 0 {
      identity = "source:\(sourceMessageId)"
    } else if !isBlank(taskId) {
      identity = "task:\(trim(taskId))"
    } else {
      return ""
    }
    return "assistant-final:\(identity)"
  }

  static func resolveTurnId(
    explicitTurnId: String,
    taskId: String,
    turnIdForTask: (String) -> String?
  ) -> String {
    let explicit = trim(explicitTurnId)
    if !explicit.isEmpty {
      return explicit
    }
    let cleanTaskId = trim(taskId)
    guard !cleanTaskId.isEmpty else { return "" }
    return trim(turnIdForTask(cleanTaskId) ?? "")
  }

  static func coalesce(_ entries: [AgentTranscriptEntry]) -> [AgentTranscriptEntry] {
    let candidates = entries.filter(isCanonicalFinalCandidate)
    guard candidates.count >= 2 else { return entries }

    let retainedIds = Set(
      Dictionary(grouping: candidates, by: duplicateKey)
        .values
        .compactMap { duplicates in
          duplicates.reduce(nil as AgentTranscriptEntry?) { best, entry in
            guard let best else { return entry }
            return isBetterCanonicalEntry(entry, than: best) ? entry : best
          }?.id
        }
    )

    return entries.filter { entry in
      !isCanonicalFinalCandidate(entry) || retainedIds.contains(entry.id)
    }
  }

  private static func isCanonicalFinalCandidate(_ entry: AgentTranscriptEntry) -> Bool {
    entry.role == .assistant &&
      entry.dedupeKey.hasPrefix("assistant-final:") &&
      !isBlank(entry.taskId) &&
      !isBlank(entry.text)
  }

  private static func duplicateKey(_ entry: AgentTranscriptEntry) -> String {
    [
      entry.conversationId,
      trim(entry.taskId),
      trim(entry.text)
    ].joined(separator: "\u{001f}")
  }

  private static func isBetterCanonicalEntry(
    _ candidate: AgentTranscriptEntry,
    than current: AgentTranscriptEntry
  ) -> Bool {
    let candidateScore = canonicalScore(candidate)
    let currentScore = canonicalScore(current)
    if candidateScore != currentScore {
      return candidateScore.lexicographicallyPrecedes(currentScore) == false
    }
    return false
  }

  private static func canonicalScore(_ entry: AgentTranscriptEntry) -> [Int64] {
    [
      isBlank(entry.turnId) ? 0 : 1,
      isBlank(entry.richOutputJson) ? 0 : 1,
      entry.timestampMillis
    ]
  }

  private static func trim(_ value: String) -> String {
    value.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private static func isBlank(_ value: String) -> Bool {
    trim(value).isEmpty
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentFinalResponseIdentity.kt
package com.galaxyssi.chat

internal object AgentFinalResponseIdentity {
    fun dedupeKey(
        turnId: String,
        sourceMessageId: Long = 0L,
        taskId: String = ""
    ): String {
        val identity = when {
            turnId.isNotBlank() -> "turn:${turnId.trim()}"
            sourceMessageId > 0L -> "source:$sourceMessageId"
            taskId.isNotBlank() -> "task:${taskId.trim()}"
            else -> return ""
        }
        return "assistant-final:$identity"
    }

    fun resolveTurnId(
        explicitTurnId: String,
        taskId: String,
        turnIdForTask: (String) -> String?
    ): String = explicitTurnId.trim().ifBlank {
        taskId.trim().takeIf(String::isNotBlank)
            ?.let(turnIdForTask)
            ?.trim()
            .orEmpty()
    }

    fun coalesce(entries: List<AgentTranscriptEntry>): List<AgentTranscriptEntry> {
        val candidates = entries.filter(::isCanonicalFinalCandidate)
        if (candidates.size < 2) return entries
        val retainedIds = candidates
            .groupBy(::duplicateKey)
            .values
            .mapTo(mutableSetOf()) { duplicates ->
                duplicates.maxWithOrNull(
                    compareBy<AgentTranscriptEntry>(
                        { it.turnId.isNotBlank() },
                        { it.richOutputJson.isNotBlank() },
                        AgentTranscriptEntry::timestampMillis
                    )
                )?.id.orEmpty()
            }
        return entries.filter { entry ->
            !isCanonicalFinalCandidate(entry) || entry.id in retainedIds
        }
    }

    private fun isCanonicalFinalCandidate(entry: AgentTranscriptEntry): Boolean =
        entry.role == AgentTranscriptRole.ASSISTANT &&
            entry.dedupeKey.startsWith("assistant-final:") &&
            entry.taskId.isNotBlank() &&
            entry.text.isNotBlank()

    private fun duplicateKey(entry: AgentTranscriptEntry): String = listOf(
        entry.conversationId,
        entry.taskId.trim(),
        entry.text.trim()
    ).joinToString("\u001f")
}
